kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+In the example, a `result()` method is also added to calculate the expected output.
+
+For the example, add the following methods to `BignumAddExample`:
+
+```python
+    def arguments(self) -> List[str]:
+        return [
+            "\"{}\"".format(self.arg_a),
+            "\"{}\"".format(self.arg_b),
+            self.result()
+        ]
+
+    def result(self) -> str:
+        # Return as a quoted hexadecimal string, without leading 0x
+        return "\"{:x}\"".format(self.int_a + self.int_b)
+```
+
+`generate_function_tests()` handles the construction of class instances for each unique
+test case, and yields `test_case.TestCase` objects from each instance.
+This is a `classmethod`, which acts on the class, rather than a specific instance.
+In the example, a list of strings will be used as inputs, and a test case is generated
+for each possible combination.
+The standard module `itertools` is used for generation of unique input combinations.
+
+For the example, add the following method to `BignumAddExample`:
+
+```python
+    @classmethod
+    def generate_function_tests(cls) -> Iterator[test_case.TestCase]:
+        # Create a list of input values
+        input_values = ["", "0", "1", "123"]
+        # combinations_with_replacement generates a list of all unique
+        # combinations, including where inputs are the same i.e. ("0", "0")
+        for a_value, b_value in itertools.combinations_with_replacement(
+            input_values, 2
+        )
+            # Initialize a class instance with these values
+            class_instance = cls(a_value, b_value)
+            # Yield the TestCase object from this instance
+            yield class_instance.create_test_case()
+```
+
+### Other BaseTarget attributes and methods
+
+`description()` returns a description of the test case, in the format "`test_name`
+#`count` `case_description`" by default.
+`case_description` is an optional description for the specific test case, which may be
+explicitly passed when constructing an instance, or generated in the class.
+This can be used to add context, and in the example this will be generated by overriding
+the `description()` method.
+
+For the example, add the following method to `BignumAddExample`:
+
+```python
+    def description(self) -> str:
+        if not self.case_description:
+            self.case_description = "{} + {}".format(
+                # hex() converts to a hex string with leading 0x
+                "0 (null)" if self.arg_a == "" else hex(self.int_a),
+                "0 (null)" if self.arg_b == "" else hex(self.int_b)
+            )
+        # Call description() in the parent class
+        return super().description()
+```
+
+`dependencies` is a list of build macros, which specifies the required build config for
+the test case to run.
+This can be set as a class attribute if always required for the test function, or set
+conditionally in the constructor if dependent on the test case.
```

Comment:
This warrants an example: how do you set a dependency for one specific test case?

(Setting a dependency for a whole function is rarely useful, since that can go in the `.function` file. It's only useful when there are multiple function classes with the same function name.)